vllm/model_executor/layers/quantization/fp8.py

```diff
@@ -81,46 +101,165 @@ def create_weights(
         params_dtype: torch.dtype,
         **extra_weight_attrs,
     ):
+        del input_size, output_size
         output_size_per_partition = sum(output_partition_sizes)
+
+        layer.process_after_load = True
+        layer.logical_widths = output_partition_sizes
+
+        # WEIGHT
+        weight_dtype = (torch.float8_e4m3fn
+                        if self.quant_config.is_checkpoint_fp8_serialized else
+                        params_dtype)
         weight = Parameter(torch.empty(output_size_per_partition,
                                        input_size_per_partition,
-                                       dtype=params_dtype),
+                                       dtype=weight_dtype),
                            requires_grad=False)
         layer.register_parameter("weight", weight)
-        set_weight_attrs(weight, {"input_dim": 1, "output_dim": 0})
-        set_weight_attrs(weight, extra_weight_attrs)
+        set_weight_attrs(weight, {
+            **extra_weight_attrs,
+            "input_dim": 1,
+            "output_dim": 0,
+        })
 
-        w_scale = Parameter(
-            torch.empty(1, dtype=torch.float32),
-            requires_grad=False,
-        )
-        layer.register_parameter("weight_scaling_factor", w_scale)
+        # If checkpoint is serialized fp8, load them.
+        # Otherwise, wait until process_weights_after_loading.
+        if self.quant_config.is_checkpoint_fp8_serialized:
+            # WEIGHT SCALE
+            weight_scale = Parameter(torch.empty(len(output_partition_sizes),
+                                                 dtype=torch.float32),
+                                     requires_grad=False)
+            layer.register_parameter("weight_scale", weight_scale)
+            set_weight_attrs(
+                weight_scale, {
+                    **extra_weight_attrs,
+                    "fp8_scales_shard_indexer":
+                    self.scales_shard_indexer,
+                })
+
+            # ACTIVATION SCALE
+            if self.quant_config.activation_scheme == "static":
+                act_scale = Parameter(torch.empty(len(output_partition_sizes),
+                                                  dtype=torch.float32),
+                                      requires_grad=False)
+                layer.register_parameter("act_scale", act_scale)
+                set_weight_attrs(
+                    act_scale, {
+                        **extra_weight_attrs,
+                        "fp8_scales_shard_indexer":
+                        self.scales_shard_indexer,
+                    })
+
+    def scales_shard_indexer(
+            self, param: torch.Tensor, loaded_weight: torch.Tensor,
+            shard_id: Union[str, int]) -> Tuple[torch.Tensor, torch.Tensor]:
+        qkv_idxs = {"q": 0, "k": 1, "v": 2}
+
+        if isinstance(shard_id, int):
```

Comment:
I don't really understand this, why do we have this if statement here? Is `shard_id` currently always `str`? If yes, should we remove the `int` handling until we need it?

Reply:
For MLP (`GateUpProj`), it is an int